src/software/util/design_patterns/generic_factory.h

```diff
@@ -0,0 +1,121 @@
+     * names to a "create" function that will create and return a unique_ptr to a new
+     * concrete instance of the Generic.
+     *
+     * @return a const reference to the Generic registry
+     */
+    static const GenericRegistry<GenericType>& getRegistry();
+
+    /**
+     * Returns a list of names of all the existing Generics
+     *
+     * @return a list of names of all the existing Generics
+     */
+    static std::vector<std::string> getRegisteredGenericNames();
+
+    /**
+     * Returns a list of constructor functions for all the existing Generics
+     *
+     * @return a list of constructor functions for all the existing Generics
+     */
+    static std::vector<std::function<std::unique_ptr<GenericType>()>>
+    getRegisteredGenericConstructors();
+
+protected:
+    /**
+     * Adds a Generic to the Generic Registry
+     *
+     * @param generic_name The name of the Generic to be added
+     * @param generic_creator A "create" function that takes no arguments and will return
+     * a unique_ptr to a new instance of the specified Generic
+     */
+    static void registerGeneric(
+            std::string generic_name,
+            std::function<std::unique_ptr<GenericType>()> generic_creator);
+
+private:
+    /**
+     * Returns a reference to the Generic registry. The registry is a map of Generic names
+     * to a "create" function that will create and return a unique_ptr to a new concrete
+     * instance of the Generic, which allows the code to be aware
+     * of all the Generics that are available.
+     *
+     * This is the same as the above public getRegistry function. We need a mutable
+     * version in order to add entries to the registry. The function is private so that
+     * only this class can make the modifications. Outside sources should not have direct
+     * access to modify the registry.
+     *
+     * @return a mutable reference to the Generic registry
+     */
+    static GenericRegistry<GenericType>& getMutableRegistry();
+};
+
+/**
+ * This templated backend factory class is used by Backends that are derived from the
+ * Abstract Backend class. Its purpose is to create a Factory for the implemented Backend
+ * and automatically register the backend in the BackendFactory registry.
+ *
+ * Declaring the static variable will also cause it to be initialized at the start of the
+ * program (because it's static). This will immediately call the constructor, which adds
+ * the backend T to the BackendFactory registry. From then on, the rest of the program
+ * can use the registry to find all the Backends that are available (and register with
+ * this templated class).
+ *
+ * @tparam T The class of the Backend to be added to the registry. For example, to add a
+ * new class called MoveBackend that inherits from Backend, the following line should be
+ * added to the end of the .cpp file (without the quotations): "static
+ * TGenericFactory<MoveBackend> factory;"
+ */
+ // TODO not sure how to 'template'
```

Comment:
This should work fine, it just means that instead of writing something like `static TPlayFactory<ShootOrPassPlay> factory;` at the bottom of a play, we would write `static TGenericFactory<Play, ShootOrPassPlay> factory;`. A bit more verbose, but not too bad.